rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,160 @@
+
+-- Do some DB changes
+
+SELECT * FROM iasql_preview(); -- Because there's no distinction between apply and sync, this function name is simpler than before
+
+-- If we don't like what we saw
+
+CALL iasql_rollback(); -- Executes `iasql_sync`, then re-enables the cron
+
+-- If we like what we saw
+
+CALL iasql_commit(); -- Re-enables the cron, but also performs a blocking execution of the new logic so a `SELECT` on the newly-created records has the desired cloud IDs
+```
+
+To emulate the two One-Way automatic modes, we can take a page from Postgres' [Table and Row-Level Locking](https://www.postgresql.org/docs/15/explicit-locking.html). We can create sets of functions to put a whole table or just specific rows into either of these modes or put them back to normal operations. These functions would write to a new metadata table that would override the decision the audit log makes, which makes them pretty easy additions to the functionality while gaining significant power and flexibility for users (and could be pure SQL stored procedures that don't go through the engine's RPC path at all, since they just manipulate a config table).
+
+Naming these functions to be clearly understood will be important so "Anti-Chaos-Monkey" can't be what we go with here. ;) The current proposal for the name for the "Write-Only Automatic" style functions is "Enforce" as the user is asking IaSQL to enforce these particular rows or tables to always match the database, and the "Read-Only Automatic" style becomes "Track" as the user is asking IaSQL to track the upstream cloud. The proposed new syntax for these is:
+
+```sql
+SELECT iasql_enforce('table_name'); -- Enforce an entire table onto the cloud.
+
+SELECT iasql_enforce('table_name', 5); -- Enforce a particular row from a table onto the cloud. We need to use the database ID here because enforcement *will* cause `@cloudId` churn if it needs to recreate the resource.
+
+SELECT iasql_enforce_all(); -- Essentially makes the DB write-only
+
+SELECT iasql_drop_enforce('table_name'); -- Drop the enforcement of the table. Should it also automatically drop specific row-level enforcement is something to decide. I'm leaning towards "yes" but it could be unexpected.
+
+SELECT iasql_drop_enforce('table_name', 5); -- Drop the enforcement of a particular row. This one, at least, is unambiguous.
+
+SELECT iasql_drop_enforce_all(); -- Restores normal behavior.
+
+SELECT iasql_track('table_name'); -- Makes the table effectively read-only.
+
+SELECT iasql_track('table_name', 5); -- Makes a particular row read-only.
+
+SELECT iasql_track_all(); -- Makes the DB read-only
+
+SELECT iasql_drop_track('table_name'); -- Drops tracking for the table. Similar issue to above: should this affect row-level tracking in the same table?
+
+SELECT iasql_drop_track('table_name', 5); -- Drops tracking for a particular row.
+
+SELECT iasql_drop_track_all(); -- Restores normal behavior
+```
```

Comment:
Wow, this would really give some flexibility, but also, it is quite tricky for us? Could a user mess up using these statements and create locks or infinite loops because we need to update a table that is set as write-only?

Reply:
I figure write-only would work like the current apply and still allow mutation to include cloud generated IDs and the like during the loop?

But there is trickiness here due to mutual exclusion and what precisely should be done with the nesting nature of all/table/row rules.

Probably should expand this section and lay out the possibilities.

Reply:
yay.... how to avoid deadlocks there, and avoid being in a broken state if there is some problem with engine, network, etc?

Reply:
I haven't addressed the broken state issue, yet, will get to that next, but I consider these two "one-way modes" more power-user-type stuff, so the rough answer is "it's on the user who broke it" but I'll try to be more specific.